packet/src/pacer.rs

```diff
@@ -0,0 +1,1032 @@
+        }
+    }
+}
+
+/// A packet Pacer.
+///
+/// The pacer is responsible for ensuring correct pacing of packets onto the network at a given
+/// bitrate.
+pub trait Pacer {
+    /// Set the target bitrate to send at. This can be exceeded if required to drain excessively
+    /// long packet queues.
+    fn set_target_bitrate(&mut self, target_bitrate: Bitrate);
+
+    /// Poll for a timeout.
+    fn poll_timeout(&self) -> Option<Instant>;
+
+    /// Handle time moving forward, should be called periodically as indicated by [`Pacer::poll_timeout`].
+    fn handle_timeout(&mut self, now: Instant, iter: impl Iterator<Item = QueueState>);
+
+    /// Determines what action to take, if any.
+    ///
+    ///
+    /// ## Return values
+    ///
+    /// * [`PollOutcome::PollQueue`] indicates that the queue indicated by the contained
+    /// [`QueueState`] should be polled and the resulting packet sent.
+    /// * [`PollOutcome::Nothing`] indicates no action is required at this time.
+    fn poll_action(&mut self) -> PollOutcome;
+
+    /// Register a packet having been sent.
+    ///
+    /// **MUST** be called each time [`Pacer::poll_action`] produces [`PollOutcome::PollQueue`]
+    /// after the packet is sent.
+    fn register_send(&mut self, now: Instant, packet_size: DataSize, from: QueueId);
+}
+
+/// A unique identifier for a given packet queue. Should be immutable.
+#[derive(Debug, Default, PartialEq, Eq, Hash, Clone, Copy)]
+pub struct QueueId(u64);
+
+/// The state of a single upstream queue.
+/// The pacer manages packets across several upstream queues.
+#[derive(Debug)]
+pub struct QueueState {
+    /// The unique identifier for this queue within the session.
+    pub id: QueueId,
+    /// The total size of queued payloads within the queue.
+    pub size: DataSize,
+    /// The total number of packets in the queue.
+    /// NB: This is not a [`usize`] because it will later be used to divide a [`Duration`], for which
+    /// [`usize`] isn't implement. If the queues end up with 2^32 packets something has gone very wrong
+    /// in any case.
+    pub packet_count: u32,
+    /// The total time the packets in the queue have spent queued.
+    pub total_queue_time: Duration,
+    /// The kind of packets this queue contains.
+    pub queue_kind: PacketKind,
+    // The last time a packet was sent from this queue.
+    last_send_time: Option<Instant>,
+    // The time the packet at the front of the queue was queued.
+    leading_packet_queue_time: Option<Instant>,
+}
+
+/// The outcome of a call to [`Pacer::poll_action`].
+#[derive(Debug, Clone, Copy)]
+pub enum PollOutcome {
+    /// The caller **MUST** poll the next packet from the queue that produced the contained [`QueueState`]
+    /// and send it.
+    PollQueue(QueueId),
+    /// The caller MUST do nothing for now.
+    Nothing,
+    // TODO: Produce padding when we are sending very infrequently or if we need it to probe.
```

Comment:
Padding it's just for when sending infrequently.  It (or probing) is important for an SFU that can't just send more when ever it wants be adjusting an encoder.